sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/cryptography/BlobDecryptionPolicy.java

```diff
@@ -0,0 +1,489 @@
+     * Initializes a new instance of the {@link BlobDecryptionPolicy} class with the specified key and resolver.
+     * <p>
+     * If the generated policy is intended to be used for encryption, users are expected to provide a key at the
+     * minimum. The absence of key will cause an exception to be thrown during encryption. If the generated policy is
+     * intended to be used for decryption, users can provide a keyResolver. The client library will - 1. Invoke the key
+     * resolver if specified to get the key. 2. If resolver is not specified but a key is specified, match the key id on
+     * the key and use it.
+     *
+     * @param key An object of type {@link IKey} that is used to wrap/unwrap the content encryption key.
+     * @param keyResolver The key resolver used to select the correct key for decrypting existing blobs.
+     */
+    BlobDecryptionPolicy(IKey key, IKeyResolver keyResolver) {
+        if (key == null && keyResolver == null) {
+            throw logger.logExceptionAsError(new IllegalArgumentException("Key and KeyResolver cannot both be null"));
+        }
+        this.keyWrapper = key;
+        this.keyResolver = keyResolver;
+    }
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        // 1. Expand the range of download for decryption
+        String rangeHeader = "x-ms-range";
+        HttpHeaders requestHeaders = context.getHttpRequest().getHeaders();
+        EncryptedBlobRange encryptedRange = EncryptedBlobRange.getEncryptedBlobRangeFromHeader(
+            requestHeaders.getValue(rangeHeader));
+
+        // Assumption: Download is the only API on an encrypted client that sets x-ms-range
+        // Only set the x-ms-range header if it already exists
+        if (requestHeaders.getValue(rangeHeader) != null) {
+            requestHeaders.put(rangeHeader, encryptedRange.toBlobRange().toString());
+        }
+
+        // 2. Replace the body of the response with a decrypted version of the body
+        return next.process().flatMap(httpResponse -> {
+            // Assumption: Download is the only API on an encrypted client that is a get request and has a body in the
+            // response
+            if (httpResponse.getRequest().getHttpMethod() == HttpMethod.GET && httpResponse.getBody() != null) {
+                HttpHeaders responseHeaders = httpResponse.getHeaders();
+                /*
+                We will need to know the total size of the data to know when to finalize the decryption. If it was
+                not set originally with the intent of downloading the whole blob, update it here.
+                 */
+                encryptedRange.withAdjustedDownloadCount(Long.parseLong(responseHeaders.getValue("Content-Length")));
+                boolean padding = encryptedRange.toBlobRange().getOffset()
+                    + encryptedRange.toBlobRange().getCount() > (blobSize(responseHeaders) - ENCRYPTION_BLOCK_SIZE);
```

Comment:
Can you comment on this line? I think it's that we expect padding only if we are at the end of the blob